build/ci_utils/src/actions/artifacts.rs

```diff
@@ -67,36 +68,47 @@ pub fn discover_recursive(
     rx.into_stream()
 }
 
-pub async fn upload(
+
+pub fn upload(
     file_provider: impl Stream<Item = FileToUpload> + Send + 'static,
-    artifact_name: impl AsRef<str>,
+    artifact_name: impl Into<String>,
     options: UploadOptions,
-) -> Result {
-    let handler =
-        ArtifactUploader::new(SessionClient::new_from_env()?, artifact_name.as_ref()).await?;
-    let result = handler.upload_artifact_to_file_container(file_provider, &options).await;
-    // We want to patch size even if there were some failures.
-    handler.patch_artifact_size().await?;
-    result
+) -> BoxFuture<'static, Result> {
+    let artifact_name = artifact_name.into();
+    let span = info_span!("Artifact upload", artifact_name);
+    async move {
+        let handler = ArtifactUploader::new(SessionClient::new_from_env()?, artifact_name).await?;
+        let result = handler.upload_artifact_to_file_container(file_provider, &options).await;
+        // We want to patch size even if there were some failures.
+        handler.patch_artifact_size().await?;
+        result
+    }
+    .instrument(span)
+    .boxed()
```

Comment:
Why `.boxed()` instead of `.await`? Also, this whole change could probably be an `#[instrument]` annotation instead.

Reply:
This is a convention to avoid unnecessary lifetime dependencies on futures — we don't want to force caller to keep arguments alive while the artifacts are uploading. To avoid spurious lifetime dependencies we avoid using `async fn` here, which in turn makes it impossible to use `#[instrument]`. 

See https://github.com/rust-lang/rust/issues/42940